# Mathematical Operations

## Fokus Bab

Bab ini membahas kumpulan fungsi matematis dasar di NumPy yang menjadi bahan baku transformasi numerik dalam praktik Data Science. Berbeda dari bab-bab sebelumnya yang berfokus pada struktur dan manipulasi array, bab ini berfokus pada apa yang dilakukan terhadap nilai di dalam array — mulai dari operasi sederhana seperti nilai absolut, hingga fungsi yang menjadi fondasi algoritma statistik dan machine learning seperti eksponensial dan logaritma.

## Tujuan Pembelajaran

Setelah menyelesaikan bab ini, Anda diharapkan mampu:

- Menerapkan fungsi matematis dasar (np.abs(), np.sqrt(), np.square(), np.exp(), np.log(), np.log10()) pada array.
- Melakukan pembulatan angka dengan np.round(), np.floor(), dan np.ceil(), serta menjelaskan perbedaan perilakunya.
- Membedakan fungsi element-wise np.maximum()/np.minimum() dengan fungsi aggregation np.max()/np.min() yang sudah dipelajari di bab 6.
- Menjelaskan hubungan setiap transformasi numerik (log, square, absolute, clipping) dengan aplikasinya yang konkret dalam Data Science dan Machine Learning.

## Isi Materi

- Fungsi Matematis Dasar — np.abs(), np.sqrt(), np.square(), np.exp(), np.log(), np.log10().
- Rounding — np.round(), np.floor(), np.ceil(), beserta perilaku pembulatan yang sering menjadi jebakan pemula.
- Element-wise Min/Max — np.maximum(), np.minimum(), dan perbedaannya dengan fungsi aggregation.
- Transformasi Data & Hubungannya dengan DS — log transformation, squaring, absolute value, clipping, serta kaitannya dengan skewness, squared error, dan ML modeling.

## Goals

Mampu melakukan transformasi numerik menggunakan NumPy.

# 1. Fungsi Matematis Dasar

## 1.1 Absolute Value & Square Root

### A. np.abs() — Absolute Value

* `np.abs()` mengembalikan nilai absolute (jarak dari nol selalu positif) dari setiap elemen.
 > Hasilnya tidak pernah neagif

In [1]:
import numpy as np

In [2]:
arr = np.array([-1, 4, -7, 9, 4, -5])
np.abs(arr)

array([1, 4, 7, 9, 4, 5])

Penjelasan kode di atas:

* `np.abs()` digunakan untuk mendapatkan **nilai absolut** dari setiap elemen array.
* Nilai negatif akan kehilangan tanda `-`, sedangkan nilai positif tetap tidak berubah.
* Operasi dilakukan secara **element-wise**, sehingga setiap elemen diproses secara independen.
* Nilai absolut menunjukkan **besar atau magnitudo** suatu nilai tanpa memperhatikan arahnya.
* Contoh:

  * `|-3| = 3`
  * `|4| = 4`
  * `|-5| = 5`
* Dalam Data Science, `np.abs()` dapat digunakan untuk menghitung **besar kesalahan prediksi**.
* Kesalahan prediksi dapat bernilai positif maupun negatif, sehingga tanda kesalahan perlu diabaikan ketika yang ingin diukur adalah **seberapa besar kesalahannya**.
* Konsep ini menjadi dasar dari **MAE (Mean Absolute Error)**, yang menghitung rata-rata nilai absolut dari seluruh error.


Secara konseptual

```text
[-3,   4,  -5,   7,  -1]
  ↓    ↓    ↓     ↓     ↓
 abs  abs  abs   abs   abs
  ↓    ↓    ↓     ↓     ↓
[ 3,   4,   5,   7,    1]
```

> tidak perlu menggunkan loops

### B. np.sqrt() — Square Root

* `np.sqrt()` untuk menghitung akar kuadrat setiap elemen array

In [3]:
arr = np.array([4, 9, 16, 25])

np.sqrt(arr)

array([2., 3., 4., 5.])

Penjelasan kode di atas dan catatan penting:

* Hasil `np.sqrt()` selalu bertipe float, meskipun input berupa bilangan bulat, karena akar kuadrat umumnya tidak menghasilkan bilangan bulat.
* Perhatian: np.sqrt() tidak mendukung nilai negatif secara default. `np.sqrt(np.array([-4]))` akan menghasilkan `nan (Not a Number)` disertai peringatan ` RuntimeWarning: invalid value encountered in sqrt`, karena akar kuadrat dari bilangan negatif tidak terdefinisi dalam ranah bilangan riil.

In [4]:
arr = np.array([-4])

np.sqrt(arr)

C:\Users\ACER\AppData\Local\Temp\ipykernel_11972\3787878387.py:3: RuntimeWarning: invalid value encountered in sqrt
  np.sqrt(arr)


array([nan])

## 1.2 Square

* `np.square()` untuk menghitung pangakt dua dari elemen array.
* Operasi dilakukan secara element-wise, elemen dikuadratkan secara independent.

In [5]:
arr = np.array([2, -3, 4])
np.square(arr)

array([ 4,  9, 16])

* Bilangan negatif menghasilkan nilai positif setelah dikuadratkan karena:
    - (-3) × (-3) = 9
    - Perkalian dua bilangan negatif menghasilkan bilangan positif.
* Sifat tersebut membuat np.square() dapat digunakan untuk menghilangkan arah/tanda error sekaligus mempertahankan besarnya error.
* Berbeda dengan np.abs(), squaring memberikan penalti yang semakin besar terhadap nilai yang jauh dari nol.

## 1.3 Exponential & Logarithm

### 1.3.1 Konsep Eksponen

Eksponen adalah operasi matematika yang menunjukkan berapa kali suatu bilangan (basis) dikalikan dengan dirinya sendiri.

Bentuk umum eksponen adalah $a^x$, dengan:
- $a$ = base/basis
- $x$ = exponent/pangkat

Contoh
- $2^3 = 2 \times 2 \times 2 = 8$
- $5^2 = 5 \times 5 = 25$

Beberapa aturan dasar eksponen:
- $a^0 = 1$
- $a^1 = a$
- $a^{-x} = \frac{1}{a^x}$
- $a^m \times a^n = a^{m+n}$
- $\frac{a^m}{a^n} = a^{m-n}$

Eksponen penting dalam Data Science karena digunakan untuk merepresentasikan pertumbuhan, probabilitas, transformasi data, dan berbagai fungsi dalam Machine Learning.

### 1.3.2 Konstanta Euler ($e$)


* $e$ adalah konstanta matematika
* Nilainya kira kira
    - $e \approx 2.718281828$
* Konstanta $e$  merupakan basis dari natural   logarithm
* Eksponen berbasi $e$ dituli sebagai
    - $e^x$

* Konsep ini menjadi dasar dari fungsi `np.exp()` di NumPy.

* $e$ banyak digunakan pada:

  * exponential growth
  * probabilitas
  * distribusi statistik
  * sigmoid
  * log-likelihood
  * neural network
  * optimisasi


### 1.3.3 `np.exp()`


* `np.exp()` digunakan untuk mengitung eksponen berbasis $e$
* Secara matematis
    -  $\text{np.exp}(x) = e^x$

* Operasi dilakukan secara element-wise pada setiap element array.

In [1]:
import numpy as np

arr = np.array([0, 1, 2])
np.exp(arr)

array([1.        , 2.71828183, 7.3890561 ])

* Karena:

  $e^0 = 1$

  $e^1 \approx 2.718$

  $e^2 \approx 7.389$

* Untuk nilai negatif:

  $e^{-x} = \frac{1}{e^x}$

* Artinya, semakin negatif nilai $x$, semakin kecil nilai hasil `np.exp(x)`.

### 1.3.4 Konsep Logaritma


* Logaritma merupakan kebalikan dari ekponensial
* Jika

    $a^x = b$

    maka

    $\log_a(b) = x$

* Dengan kata lain logaritma menjawab pertanyaan
    > "Pangkat berapa yang harus diberikan kepada basis $a$ agar mengasilkan $b$?"

* Contoh

    $2^3 = 8$

    sehingga

    $\log_2(8) = 3$

* Hubungan eksponen dari logaritma:

    $a^x = b \iff \log_a(b) = x$

* Konsep inverse ini sangat penting dalam Data Science karena memungkinkan transformasi dari skala eksponensial kembali ke skala yang lebih mudah dianalisis.

### 1.3.5 Basis Logaritma


Terdapat beberapa basis logaritma yang umum digunakan.



#### Natural Logarithm

* Natural logarithm menggunakan basis $e$.

* Ditulis sebagai:

  $\ln(x) = \log_e(x)$

* Di NumPy, natural logarithm dihitung menggunakan:

    ```python
    np.log(x)
    ```

* Penting untuk diperhatikan bahwa `np.log()` berarti natural logarithm, bukan logaritma basis 10.

#### Common Logarithm

* Common logarithm menggunakan basis 10.

* Ditulis sebagai:

  $\log_{10}(x)$

* Di NumPy:

    ```python
    np.log10(x)
    ```

### 1.3.6 `np.log()`

* `np.log()` digunakan untuk menghitung natural logarithm.

* Secara matematis:

  $\text{np.log}(x) = \ln(x) = \log_e(x)$

* Operasi dilakukan element-wise.

In [3]:
import numpy as np

arr = np.array([1, np.e, 10, 100])

print(arr)
np.log(arr)

[  1.           2.71828183  10.         100.        ]


array([0.        , 1.        , 2.30258509, 4.60517019])

* Hal ini terjadi karena:

  $\ln(1) = 0$

  $\ln(e) = 1$

  $\ln(10) \approx 2.3026$

  $\ln(100) \approx 4.6052$

* `np.log()` banyak digunakan dalam:

  * log transformation
  * log-likelihood
  * Maximum Likelihood Estimation
  * cross-entropy
  * probabilistic modeling
  * numerical computation
